crates/doc/src/ser.rs

```diff
@@ -75,16 +92,21 @@ pub struct SerOwned<'p> {
     policy: &'p SerPolicy,
 }
 
-impl<'p, N: AsNode> serde::Serialize for SerNode<'p, N> {
+impl<'p, 'n, N: AsNode> serde::Serialize for SerNode<'p, 'n, N> {
     fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
     where
         S: ::serde::Serializer,
     {
         match self.node.as_node() {
-            Node::Array(arr) => serializer.collect_seq(arr.iter().map(|d| Self {
-                node: d,
-                policy: self.policy,
-            })),
+            Node::Array(arr) => {
+                let child_policy = self.policy.for_child();
+                serializer.collect_seq(arr.iter().take(self.policy.array_truncate_after).map(|d| {
+                    SerNode {
```

Comment:
I'm somewhat concerned about performance here, as per profiling we do spend a good amount of time serializing as we move data into Store requests.

Rather than pushing a new cloned policy onto the stack at each level, could you instead track depth on SerNode? What about making SerNode a const generic over a boolean which indicates if it's the root or not, which would let you do a compile time branch when evaluating the policy? This would be zero cost -- no extra stack allocation or runtime branches.

Or, slightly fancier, a const generic integer that does a bounded degree of compile-time depth tracking, by having branches for depths 0, 1, 2, .. X, where depth isn't incremented beyond X. Right now we don't need more than 0 (root) and 1 (node).

Reply:
a concrete refinement after thinking on it a bit more: it could be generic over const `Depth: usize`, which is instantiated today only with values `0` and `usize::MAX`. In the future, if needed, we could introduce depths `1`, `2`, etc.

Reply:
I see you added PolicyHelper but IMO it's more verbose and indirected as compared to ☝️ , and also depends on an `obj_truncate_after` concept which we don't have an expected use case for and IMO ought to remove, lest some future reader is lead to believe it _is_ required and must thus be supported indefinitely. I continue to think a Depth const generic makes more sense.

Reply:
I could see adding a non-generic `depth: usize` property to the struct, but I don't think it'd work to do something like `struct SerNode<const Depth: usize, ...>` (assuming I'm understanding your suggestion right). The problem I see with that is that the depth can't be `const`, since we only know it by recursively adding one to the previous value. Do you know of a way to make that work?

Reply:
Just to clarify, I'm reasonably sure that stable rust won't permit us to use const generics in this way, but am open to being proven wrong. Adding a `depth: usize` field to `SerNode` seems like it could work, though. TBH this seems to me like an implementation detail that'd be easy to change, so I'm inclined to leave it as is until we can see what the next requirements are. But can refactor to use the `depth: usize` approach if you feel strongly about it.

Reply:
> since we only know it by recursively adding one to the previous value. Do you know of a way to make that work?

Yea, you can't do recursive Depth + 1 -- the recursive addition breaks -- but you _can_ do:

```
if Depth == 0 {
   SerPolicy::<1>::something()
} else if Depth == 1 {
   SerPolicy::<2>::something()
 } else if Depth == 2 {
   SerPolicy::<3>::something() // and so on...
} else {
   SerPolicy::<usize::MAX>::something()
}
```

And just unroll it only for the actual levels of depth where you plan to have differentiated behavior (which is just Depth = 0 vs Depth = usize::MAX here, so just one branch for now)